KAFKA-14793: Propagate Topic Ids to the Group Coordinator during Offsets Commit

Hangleton opened the pull request:
[KAFKA-14793 - Propagate Topic Ids to the Group Coordinator during Offsets Commit](https://issues.apache.org/jira/browse/KAFKA-14793)

This PR is a pre-requisite for [PR-13240](https://github.com/apache/kafka/pull/13240). It introduces topic ids in the signature of the group coordinator `commitOffsets` method, and performs a conversion between `TopicPartition`  and `TopicIdPartition` when interacting with the group metadata manager. Topic ids are not propagated to that component yet. 

In this PR, topic ids are resolved from the topic names provided in the offset commit request. With [PR-13240](https://github.com/apache/kafka/pull/13240), ids will be provided by the request from version 9.

Eventually, topic ids will be part of the data model authored by the group coordinator, but the corresponding change is out of scope of this PR.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Hangleton committed:
Adopt TopicPartition to/from TopicIdPartition in the group coordinator to prepare the introduction of topic IDs in the Offset Commit API.

dajac commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -1034,16 +1035,21 @@ private[group] class GroupCoordinator(
       if (validationErrorOpt.isDefined) {
         responseCallback(offsetMetadata.map { case (k, _) => k -> validationErrorOpt.get })
       } else {
+        // GroupMetadataManager does not support topic ids yet.
```
I don't really see the need for `TopicIdAdaptor`. Could we just update `storeOffsets` to take `TopicIdPartition` as well? Updating the method signatures is fine but we should not update the underlying Map where offsets are stored. Am I missing something?

Hangleton replied:
You are right, it is simpler to propagate one more layer until the final data are constructed. Thanks, updating the PR.

dajac commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinatorAdapter.scala:
```diff
@@ -388,18 +393,27 @@ private[group] class GroupCoordinatorAdapter(
-          partition.commitTimestamp,
-          expireTimeMs
-        )
+      val topicId = topicIds.get(topic.name)
+      if (Uuid.ZERO_UUID.equals(topicId)) {
+        val topicPartitions = topic.partitions.asScala.map { p =>
+          new TopicPartition(topic.name, p.partitionIndex) -> Errors.UNKNOWN_TOPIC_OR_PARTITION
+        }
+        appendToResponse(topicPartitions.toMap)
+      } else {
```
This is not needed as well as it will be done in the KafkaApis layer. We should just take the topic id that we receive here and use it to create the `TopicIdPartition`. Obviously, it will always be `ZERO_UUID` at the moment.

Hangleton replied:
I see, I thought we would have to use the resolved id instead of the zero topic id. This changes the approach. Thanks for the quick check, I will fix this.

Hangleton committed:
Fix horrible mistakes.

Hangleton commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -971,8 +971,11 @@ private[group] class GroupCoordinator(
       } else {
-        groupManager.storeOffsets(group, memberId, offsetMetadata, responseCallback, producerId,
-          producerEpoch, requestLocal)
+        val offsets = offsetMetadata.map { case(k,v) => new TopicIdPartition(Uuid.ZERO_UUID, k) -> v }
```
Please note that the signature of the offset commits for transactions is not changed, hence these conversions. It is possible to increase the scope to cover transaction commit offsets but erring on the side of caution for this PR.

dajac replied:
I suppose that we can remove this now, isn't it?

Hangleton commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -469,11 +469,11 @@ class GroupMetadataManager(brokerId: Int,
           } else {
             group.inLock {
-              group.prepareOffsetCommit(offsetMetadata)
+              group.prepareOffsetCommit(offsetMetadata.map { case(k, v) => k.topicPartition -> v })
```
I didn't propagate the topic ids into `GroupMetadata` where the actual offset maps are. So, the conversion of keys from `TopicIdPartition` to `TopicPartition` happens here for both offset commit and transaction offset commit. Please let me know if we should go one level deeper and do this conversion in `GroupMetadata` itself (without changing the retained data).

dajac replied:
Would it make sense to go one level deeper?

Hangleton replied:
Sure - I pushed a new commit [b6d9d2b](https://github.com/apache/kafka/pull/13378/commits/b6d9d2ba247a7880002486b23bd86db63c03212e) which propagates `TopicIdPartition` to the following `GroupMetadata` methods:

```
onOffsetCommitAppend
prepareOffsetCommit
failPendingOffsetWrite
```

and:

```
onTxnOffsetCommitAppend
prepareTxnOffsetCommit
failPendingTxnOffsetCommit
```

Please note that not all public methods of `GroupMetadata` use `TopicIdPartition`, and some of them remain `TopicPartition`-based. Do you think it is fine? (since `GroupMetadata` is an internal class, it should be I suppose).

Hangleton replied:
Alternatively, a forward-looking approach would be to update all public methods of `GroupMetadata` to `TopicIdPartition` although that would mean drastically extending the scope to all callers beyond the offset commit path and using the zero topic id where required. I guess this wouldn't be advisable though?

dajac replied:
I would only change the ones related to the offset commit path (transactional and non-transactional) for now. Later one, I think that we may have to change the ones related to the fetch path as well because we will have the same issues.

Hangleton replied:
Makes sense!

Hangleton commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorConcurrencyTest.scala:
```diff
@@ -295,8 +295,12 @@ class GroupCoordinatorConcurrencyTest extends AbstractCoordinatorConcurrencyTest
 
-  class CommitTxnOffsetsOperation(lock: Option[Lock] = None) extends CommitOffsetsOperation {
-    override def runWithCallback(member: GroupMember, responseCallback: CommitOffsetCallback): Unit = {
+  class CommitTxnOffsetsOperation(lock: Option[Lock] = None) extends GroupOperation[TxnCommitOffsetCallbackParams, TxnCommitOffsetCallback] {
```
Changed to subclass `GroupOperation` directly to use the different sets of type parameters `TxnCommitOffsetCallbackParams` and `TxnCommitOffsetCallback` which remains topic-name-based unlike their modified counterparts `CommitOffsetCallbackParams` and `CommitOffsetCallback`. This could be avoided if the signature change is extended to transaction offset commit as well.

dajac replied:
Are there any downsides of updating the transactional offset commit as well? I can't think of any.

Hangleton replied:
Cannot think of one either. Have extended to transaction offset commits in the latest git commit. Now focusing on tests.

Hangleton commented:
Apologies for the delay, David - I have applied the changes to correct the PR. There may still be under-propagation of topic ids, though. The class `GroupMetadata` is intentionally left unchanged. The PR is not yet completed because I need to review the tests and ensure we have proper coverage, although there exists coverage already from the existing updated tests. I will aim to set the PR ready for review on Monday by EOD.

Hangleton committed:
Propagate TopicIdPartitions to GroupMetadata methods called on the offset commit path. Internal mappings remain by TopicPartitions.

Hangleton committed:
Update GroupCoordinator[Adapter] transaction offset commits method to use TopicIdPartition instead of TopicPartition.

Hangleton committed:
Add assertions for GroupMetadata tests to cover access by TopicIdPartition.

dajac commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinatorAdapter.scala:
```diff
@@ -388,10 +388,10 @@ private[group] class GroupCoordinatorAdapter(
     request.topics.forEach { topic =>
       topic.partitions.forEach { partition =>
-        val tp = new TopicPartition(topic.name, partition.partitionIndex)
+        val tp = new TopicIdPartition(Uuid.ZERO_UUID, new TopicPartition(topic.name, partition.partitionIndex))
```
nit: There is another constructor which takes the uuid, partition id and topic name. Could we use that one?

dajac commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -649,26 +649,27 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+  def prepareOffsetCommit(offsets: Map[TopicIdPartition, OffsetAndMetadata]): Unit = {
     receivedConsumerOffsetCommits = true
-    pendingOffsetCommits ++= offsets
+    pendingOffsetCommits ++= offsets.map { case(k, v) => k.topicPartition -> v }
```
nit: I wonder what is best here: using ++= (addAll) or iterating over offsets and use +=. The latter does not create an intermediate collection. What do you think?
nit2: `case (k, v)`. We put a space after `case`.

Hangleton replied:
That makes sense, I suppose this is avoidable GC activity.

dajac commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -824,9 +827,18 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
 
   def offset(topicPartition: TopicPartition): Option[OffsetAndMetadata] = offsets.get(topicPartition).map(_.offsetAndMetadata)
 
+
```
nit: This can be reverted.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala:
```diff
@@ -636,6 +636,7 @@ class GroupCoordinatorAdapterTest {
   @ApiKeyVersionsSource(apiKey = ApiKeys.OFFSET_COMMIT)
   def testCommitOffsets(version: Short): Unit = {
     val groupCoordinator = mock(classOf[GroupCoordinator])
+
```
nit: This can be reverted.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorConcurrencyTest.scala:
```diff
@@ -298,13 +299,14 @@ class GroupCoordinatorConcurrencyTest extends AbstractCoordinatorConcurrencyTest
   class CommitTxnOffsetsOperation(lock: Option[Lock] = None) extends CommitOffsetsOperation {
     override def runWithCallback(member: GroupMember, responseCallback: CommitOffsetCallback): Unit = {
       val tp = new TopicPartition("topic", 0)
-      val offsets = immutable.Map(tp -> OffsetAndMetadata(1, "", Time.SYSTEM.milliseconds()))
+      val topicId = Uuid.randomUuid()
```
nit: Let's combine these two lines in `new TopicIdPartition...`. Then we can reuse it at L303 as well.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:
```diff
@@ -160,8 +160,8 @@ class GroupCoordinatorTest {
     // OffsetCommit
-    val topicPartition = new TopicPartition("foo", 0)
-    var offsetCommitErrors = Map.empty[TopicPartition, Errors]
+    val topicPartition = new TopicIdPartition(Uuid.randomUuid(), new TopicPartition("foo", 0))
```
nit: Use the second constructor. There are other cases in this file.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:
```diff
@@ -2559,10 +2559,11 @@ class GroupCoordinatorTest {
 
     // The simple offset commit should now fail
     val tp = new TopicPartition("topic", 0)
+    val tip = new TopicIdPartition(Uuid.randomUuid(), tp)
```
nit: Let's combine them.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:
```diff
@@ -2571,15 +2572,16 @@ class GroupCoordinatorTest {
   @Test
   def testFetchOffsets(): Unit = {
+    val tip = new TopicIdPartition(Uuid.randomUuid(), new TopicPartition("topic", 0))
     val tp = new TopicPartition("topic", 0)
```
nit: Ditto. There are other cases in this file.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:
```diff
@@ -2768,7 +2776,10 @@ class GroupCoordinatorTest {
-    val partitions = List(new TopicPartition("topic1", 0), new TopicPartition("topic2", 0))
+    val partitions = List(
+      new TopicIdPartition(Uuid.randomUuid(), new TopicPartition("topic1", 0)),
+      new TopicIdPartition(Uuid.randomUuid(),  new TopicPartition("topic2", 0))
```
nit: There is an extra space after `,`.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala:
```diff
@@ -1249,18 +1249,19 @@ class GroupMetadataManagerTest {
   def testCommitOffset(): Unit = {
     val memberId = ""
     val topicPartition = new TopicPartition("foo", 0)
+    val topicIdPartition = new TopicIdPartition(Uuid.randomUuid(), topicPartition)
```
nit: Let's combine them.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataTest.scala:
```diff
@@ -386,175 +386,269 @@ class GroupMetadataTest {
     group.prepareOffsetCommit(Map(partition -> offset))
-    assertTrue(group.hasOffsets)
-    assertEquals(None, group.offset(partition))
+    assertOffsets(
```
What's the reason for changing all those assertions here? It is usually better to avoid doing double refactoring like this in the same PR because there is a risk of introducing issues in the tests as well. I would rather prefer to keep assertions as they were before so I know that they are good to cover the refactoring that we are doing to introduce the topic id. Note that we could do this refactoring in a separate PR if you believe that it is necessary.

Hangleton replied:
Sure, that makes sense. This was just an attempt to improve readability by avoiding the sequences of added `assertEquals`, but I understand the concern.

dajac reviewed (commented):
Thanks for the PR. I left a few small comments.

Hangleton committed:
Minor: directly use map of TopicIdPartition -> Error/Status instead of assertion on looked-up value.

Hangleton commented:
Thank you David for the continued review, updating the PR to address the review comments.

Hangleton committed:
Use explicit assertions on committed/pending commit/tx pending commit offsets in GroupMetadataTest.

Hangleton committed:
Refactor tests to simplify TopicIdPartition declarations. Do not create an intermediate data structure when appending new offsets to pending offset commits.

Hangleton commented:
Thanks David for the review, all comments should be addressed with the new commits. I wonder how useful it would be to extend the tests in `GroupCoordinatorTest` to exercise the operations on one `TopicIdPartition` while another topic id is already registered?

dajac commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -649,26 +650,27 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+  def prepareOffsetCommit(offsets: Map[TopicIdPartition, OffsetAndMetadata]): Unit = {
     receivedConsumerOffsetCommits = true
-    pendingOffsetCommits ++= offsets
+    offsets.foreach { case (k,v) => pendingOffsetCommits += k.topicPartition -> v }
```
nit: We could use `forKeyValue` here. Moreover, let's use `(topicIdPartition, offsetAndMetadata)` (note the space after the `,`). instead of `(k, v)`.

dajac commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -827,6 +831,14 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
   // visible for testing
   private[group] def offsetWithRecordMetadata(topicPartition: TopicPartition): Option[CommitRecordMetadataAndOffset] = offsets.get(topicPartition)
 
+  private[group] def pendingOffsetCommit(topicIdPartition: TopicIdPartition): Option[OffsetAndMetadata] = {
```
Are those two methods for testing? If so, could we add a comment for each: `Used for testing`.

Hangleton replied:
Yes, they are used for testing exclusively.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorConcurrencyTest.scala:
```diff
@@ -284,7 +284,8 @@ class GroupCoordinatorConcurrencyTest extends AbstractCoordinatorConcurrencyTest
     }
     override def runWithCallback(member: GroupMember, responseCallback: CommitOffsetCallback): Unit = {
       val tp = new TopicPartition("topic", 0)
-      val offsets = immutable.Map(tp -> OffsetAndMetadata(1, "", Time.SYSTEM.milliseconds()))
+      val topicId = Uuid.randomUuid()
```
nit: Could we combine those two into on `TopicIdPartition`?

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:
```diff
@@ -2768,7 +2768,10 @@ class GroupCoordinatorTest {
     // Marker for only one partition is received. That commit should be materialized while the other should not.
 
-    val partitions = List(new TopicPartition("topic1", 0), new TopicPartition("topic2", 0))
+    val partitions = List(
```
nit: `topicIdPartitions`?

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:
```diff
@@ -2847,7 +2851,10 @@ class GroupCoordinatorTest {
     // Each partition's offsets should be materialized when the corresponding producer's marker is received.
 
-    val partitions = List(new TopicPartition("topic1", 0), new TopicPartition("topic2", 0))
+    val partitions = List(
```
ditto.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorTest.scala:
```diff
@@ -2917,34 +2925,33 @@ class GroupCoordinatorTest {
     val offset1 = offsetAndMetadata(15)
     val offset2 = offsetAndMetadata(16)
     val offset3 = offsetAndMetadata(17)
+    val (topicId, otherTopicId) = (Uuid.randomUuid(), Uuid.randomUuid())
+    val tip1 = new TopicIdPartition(topicId, 0, "topic")
+    val tip2 = new TopicIdPartition(topicId, 1, "topic")
+    val tip3 = new TopicIdPartition(otherTopicId, 0, "other-topic")
```
nit: Could we keep this block first as it was? Moreover, I would define it as follow:
```
    val tip1 = new TopicIdPartition(Uuid.randomUuid(), 0, "topic")
    val tip2 = new TopicIdPartition(tip1.topic(), 1, "topic")
    val tip3 = new TopicIdPartition(tip1, 0, "other-topic")
```

Hangleton replied:
Thanks, applied the same pattern in few other places too.

dajac commented on core/src/test/scala/unit/kafka/coordinator/group/GroupMetadataManagerTest.scala:
```diff
@@ -1824,9 +1854,13 @@ class GroupMetadataManagerTest {
     val clientHost = "localhost"
     val topic = "foo"
+    val topicId = Uuid.randomUuid()
     val topicPartition1 = new TopicPartition(topic, 0)
     val topicPartition2 = new TopicPartition(topic, 1)
     val topicPartition3 = new TopicPartition(topic, 2)
+    val topicIdPartition1 = new TopicIdPartition(topicId, topicPartition1)
+    val topicIdPartition2 = new TopicIdPartition(topicId, topicPartition2)
+    val topicIdPartition3 = new TopicIdPartition(topicId, topicPartition3)
```
nit: Could we combine them here as well?

dajac reviewed (commented):
@Hangleton Thanks for the update. I left a few nits. We should be good to go once they are addressed.

Hangleton commented:
Thanks David for the review, addressed the comments and inlined the declaration of topic-id-partitions in the locations above and in few other remaining places. The PR has just been rebased.

dajac reviewed (approved):
LGTM, thanks for the patch!

Hangleton commented:
The following tests (seems unrelated to this PR) are failing:

```
org.apache.kafka.common.network.SelectorTest.testConnectionsByClientMetric()
kafka.zk.ZkMigrationIntegrationTest.[1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT
```

Is this expected?

Hangleton commented:
`testConnectionsByClientMetric()` in `SelectorTest` failed because the client connection could not be established before the test timeout. What about introducing a connection timeout and a retry in the `PlaintextSender` from where the connection attempt is made?

mumrah commented:
@Hangleton, yes we should expect some improvement to ZkMigrationIntegrationTest with that PR. There are some cases where we lose the ZK session during the tests which causes a spurious failure. Yesterday, we identified another source of flakiness in that test which turns out to be a real bug https://issues.apache.org/jira/browse/KAFKA-14805

Hangleton commented:
Thanks David @mumrah for the follow-up!

dajac commented:
That's weird. The build is still reported `in progress` in the PR but the build has completed. The last build results are here: https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-13378/21/tests. Here are the failed tests:
```
Build / JDK 11 and Scala 2.13 / shouldAllowRemovingAndAddingNamedTopologyToRunningApplicationWithMultipleNodesAndResetsOffsets() – org.apache.kafka.streams.integration.NamedTopologyIntegrationTest
16s
Build / JDK 17 and Scala 2.13 / testReturnRecordsDuringRebalance() – org.apache.kafka.clients.consumer.KafkaConsumerTest
<1s
Build / JDK 17 and Scala 2.13 / testTrustStoreAlter(String).quorum=kraft – kafka.server.DynamicBrokerReconfigurationTest
12s
Build / JDK 17 and Scala 2.13 / testTaskRequestWithOldStartMsGetsUpdated() – org.apache.kafka.trogdor.coordinator.CoordinatorTest
2m 0s
Existing failures - 6
Build / JDK 11 and Scala 2.13 / [1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.zk.ZkMigrationIntegrationTest
1m 13s
Build / JDK 11 and Scala 2.13 / [1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.zk.ZkMigrationIntegrationTest
1m 25s
Build / JDK 17 and Scala 2.13 / [1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.zk.ZkMigrationIntegrationTest
1m 18s
Build / JDK 17 and Scala 2.13 / [1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.zk.ZkMigrationIntegrationTest
1m 27s
Build / JDK 8 and Scala 2.12 / [1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.zk.ZkMigrationIntegrationTest
1m 12s
Build / JDK 8 and Scala 2.12 / [1] Type=ZK, Name=testDualWrite, MetadataVersion=3.4-IV0, Security=PLAINTEXT – kafka.zk.ZkMigrationIntegrationTest
```
None of those are related to this PR. I am going to merge the PR.

Hangleton commented:
Thanks for the merge David.
